File: ProjectCS/SessionResults.cs
using System;
using System.Collections.Generic;

namespace SessionResultsCS
{
    public class SessionResults
    {
        // Список дисциплин, по которым получены результаты
        private DisciplineList _disciplines;

        // Список полученных результатов
        private List<AttestationResult> _results;


        public SessionResults()
        {
            _disciplines = new DisciplineList();
            _results = new List<AttestationResult>();
        }


        // Возвращает результат по дисциплине либо null, если по этой дисциплине результата нет
        public AttestationResult? GetResult(Discipline discipline)
        {
            int index = _disciplines.FindDiscipline(discipline);
            if (index == DisciplineList.DISCIPLINE_NOT_FOUND)
            {
                return null;
            }
            return _results[index];
        }

        // Устанавливает результат по дисциплине
        public void SetResult(Discipline discipline, AttestationResult result)
        {
            int index = _disciplines.FindDiscipline(discipline);
            if (index == DisciplineList.DISCIPLINE_NOT_FOUND)
            {
                _disciplines.AddDiscipline(discipline);
                _results.Add(result);
            }
            else
            {
                _results[index] = result;
            }
        }

        // Удаляет результат по дисциплине
        public void DeleteResult(Discipline discipline)
        {
            int index = _disciplines.FindDiscipline(discipline);
            if (index != DisciplineList.DISCIPLINE_NOT_FOUND)
            {
                _disciplines.RemoveDisciplineAt(index);
                _results.RemoveAt(index);
            }
        }

    }
}

File: Project/Data/SessionResults.cpp
#include "Data/SessionResults.hpp"


SessionResults::SessionResults(): m_disciplines(), m_results() {}

SessionResults::SessionResults(const SessionResults& other): m_disciplines(other.m_disciplines), m_results(other.m_results) {}


const DisciplineList& SessionResults::GetDisciplineList() const {
	return m_disciplines;
}

std::shared_ptr<AttestationResult> SessionResults::GetResult(std::shared_ptr<Discipline> discipline) const {
	int index = m_disciplines.FindDiscipline(discipline);
	if (index != -1) {
		return m_results[index];
	} else {
		return nullptr;
	}
}

void SessionResults::SetResult(std::shared_ptr<Discipline> discipline, std::shared_ptr<AttestationResult> result) {
	int index = m_disciplines.FindDiscipline(discipline);
	if (index != -1) {
		m_results[index] = result;
	} else {
		m_disciplines.AddDiscipline(discipline);
		m_results.push_back(result);
	}
}

void SessionResults::DeleteResult(std::shared_ptr<Discipline> discipline) {
	int index = m_disciplines.FindDiscipline(discipline);
	if (index != -1) {
		m_disciplines.RemoveDisciplineAt(index);
		m_results.erase(m_results.begin() + index);
	}
}
